### PR TITLE
feat: [PL-29915]: Allow direct access to secrets of Azure KV, AWS SM and GCP SM

### PR DESCRIPTION
**Pull request by nishantsaini20**
## Describe your changes
User will now be able to refer to secrets of azure KV, AWS SM and GCP SM using expression such as
`azurevault://My_AzureKV/mySecret`
`awssecretsmanager://My_AwsSM/mySecret`
Following expression for GCP Secret Manager assumes version as `latest`:
`gcpsecretsmanager://My_GoogleSM/mySecret`
For GCP Secret Manager specific version is supported by the following expression:
`gcpsecretsmanager://My_GoogleSM/mySecret/5`
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)

### CODE REVIEW

**Comment by nishantsaini20** [comment]
trigger messagecheck

**Comment by nishantsaini20** [comment]
trigger checkstylecodeformat

**Comment by nishantsaini20** [comment]
trigger checkstylecodeformat

**Review by piyush-harness — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)* status: Some of the checks are too old. 5 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/40406#-NI6mEkN70lTJVCGE29D:-NI6mEkOAuqCwfPpGMRs:b-m4ihxl) ([raw file](https://github.com/harness/harness-core/blob/8148ee8bbbb65dd2f8ff847b2f6922615e323836/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java#L51)):*
> ```Java
>         String.format("Fully-qualified path expression [%s] has illegal format.", secretIdentifier);
>     try {
>       encryptionTypeName = secretDetails[0];
> ```

Check for null and expected length before accessing index, Do not drive code flow via exception 
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java` line 89 at r1](https://reviewable.io/reviews/harness/harness-core/40406#-NI6mq6_5S9x8sMxIFqO:-NI6mq6_5S9x8sMxIFqP:bkwxrof) ([raw file](https://github.com/harness/harness-core/blob/8148ee8bbbb65dd2f8ff847b2f6922615e323836/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java#L89)):*
> ```Java
>       SecretRefParsedDataBuilder secretRefParsedDataBuilder, EncryptionType encryptionType, String fullPath) {
>     switch (encryptionType) {
>       case AWS_SECRETS_MANAGER:
> ```

Why empty case?

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)* status: Some of the checks are too old. 5 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java` line 89 at r1](https://reviewable.io/reviews/harness/harness-core/40406#-NI6mq6_5S9x8sMxIFqO:-NI6nKd8-QXdF8CMgYZR:b-pdcwp4) ([raw file](https://github.com/harness/harness-core/blob/8148ee8bbbb65dd2f8ff847b2f6922615e323836/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java#L89)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Why empty case?
</blockquote></details>

Logic for both encryption types is exactly same, so in both case calling same method.

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)* status: 5 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/40406#-NI6mEkN70lTJVCGE29D:-NI72dxz4ZKCPbxRUc8x:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8148ee8bbbb65dd2f8ff847b2f6922615e323836/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java#L51)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Check for null and expected length before accessing index, Do not drive code flow via exception 
</blockquote></details>

Done.

**Comment by piyush-harness** [comment]
*[`890-sm-core/src/main/java/io/harness/encryption/SecretRefParsedData.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/40406#-NIBZ99T2doiv_4XG0Y-:-NIBZ99T2doiv_4XG0Y0:bh0mc69) ([raw file](https://github.com/harness/harness-core/blob/044fcdfdb9179e82636f7605117829e0dde0ceba/890-sm-core/src/main/java/io/harness/encryption/SecretRefParsedData.java#L25)):*
> ```Java
>   private String relativePath;
>   private EncryptionType encryptionType;
>   AdditionalMetadata additionalMetadata;
> ```

private?

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)* status: 5 file(s) to review. 1 change requests. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`890-sm-core/src/main/java/io/harness/encryption/SecretRefParsedData.java` line 25 at r3](https://reviewable.io/reviews/harness/harness-core/40406#-NIBZ99T2doiv_4XG0Y-:-NICkRsD-mNCV69MpiSD:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/044fcdfdb9179e82636f7605117829e0dde0ceba/890-sm-core/src/main/java/io/harness/encryption/SecretRefParsedData.java#L25)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

private?
</blockquote></details>

Done.

**Review by piyush-harness — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 1 of 2 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)* status: 5 file(s) to review. 1 of 2 approvals obtained.

**Review by piyush-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406#-:-NIDZwHP6gcYXMl9AYtG:bnfp4nl)* status: 5 file(s) to review. 1 of 2 approvals obtained.

File: 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        break;
+      case GCP_SECRETS_MANAGER:
+        fillGCPSecretManagerDetails(secretRefParsedDataBuilder, fullPath);
+        break;
+      case VAULT:
+        fillVaultDetails(secretRefParsedDataBuilder, fullPath);
+        break;
+      default:
+        throw new UnsupportedOperationException(
+            String.format("Encryption type [%s] is not supported.", encryptionType));
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/DynamicSecretReferenceHelper.java:126]: Error msg `Encryption type [%s] is not supported in fully-qualified path expression.`

**Review by meenakshi-harness — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/40406)* status: :shipit: complete! ✓

**Pull request merged**